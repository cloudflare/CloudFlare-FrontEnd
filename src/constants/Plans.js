const LOCALIZED_PRO_PLAN_ID = 'constants.plans.pro';
const LOCALIZED_BIZ_PLAN_ID = 'constants.plans.biz';
const LOCALIZED_ENT_PLAN_ID = 'constants.plans.ent';

export const FREE_PLAN = 'free';
export const PRO_PLAN = 'pro';
export const BIZ_PLAN = 'business';
export const ENT_PLAN = 'enterprise';

export function planNeedsUpgrade(currentPlan, minimumPlan) {
  var planList = {};
  planList[FREE_PLAN] = 0;
  planList[PRO_PLAN] = 1;
  planList[BIZ_PLAN] = 2;
  planList[ENT_PLAN] = 3;

  return planList[currentPlan] < planList[minimumPlan];
}

export function getLocalizedPlanId(planName) {
  var localizedPlanName = planName;
  switch (planName) {
    case PRO_PLAN:
      localizedPlanName = LOCALIZED_PRO_PLAN_ID;
      break;
    case BIZ_PLAN:
      localizedPlanName = LOCALIZED_BIZ_PLAN_ID;
      break;
    case ENT_PLAN:
      localizedPlanName = LOCALIZED_ENT_PLAN_ID;
      break;
    default:
      // This should never happen
      break;
  }

  return localizedPlanName;
}
