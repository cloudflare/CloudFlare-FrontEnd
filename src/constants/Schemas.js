import { normalize, Schema, arrayOf } from 'normalizr';

const zoneSchema = new Schema('zones', { idAttribute: 'name' });
const zoneRailgunSchema = new Schema('railguns', { idAttribute: 'id' });

export function normalizeZoneByIdGetAll(zoneId, result) {
  var zoneSchemaById = new Schema(zoneId, { idAttribute: 'id' });
  return normalize(result, arrayOf(zoneSchemaById));
}

export function normalizeZoneGetAll(result) {
  return normalize(result, arrayOf(zoneSchema));
}

export function normalizeZoneRailgunGetAll(result) {
  return normalize(result, arrayOf(zoneRailgunSchema));
}
