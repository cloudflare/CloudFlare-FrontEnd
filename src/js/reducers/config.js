import * as ActionTypes from '../constants/ActionTypes';

const initialState = {
    config: {},
    isFetching: false
};

export const ABSOLUTE_URL_BASE_KEY = "absoluteUrlBase";

export function configReducer(state = initialState, action) {
    switch (action.type) {
        case ActionTypes.CONFIG_FETCH:
            return Object.assign({}, state, {
                isFetching: true
            });
        case ActionTypes.CONFIG_FETCH_SUCCESS:
            return Object.assign({}, state, {
                config: action.config,
                isFetching: false
            });
        case ActionTypes.CONFIG_FETCH_ERROR:
            return Object.assign({}, state, {
                isFetching: false
            });
        case ActionTypes.CONFIG_UPDATE_BY_KEY:
            return Object.assign({}, state, {
                    config: Object.assign({}, state.config, {
                        [action.key]: action.value
                    })
                }
            );
        default:
            return state;
    }
}