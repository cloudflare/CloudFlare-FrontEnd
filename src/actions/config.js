import http from 'cf-util-http';

import * as ActionTypes from '../constants/ActionTypes';
import { asyncIntlFetchTranslations } from './intl';
import { notificationAddError } from './notifications';
import { isLoggedIn, getEmail } from '../utils/Auth/Auth';
import { asyncUserLoginSuccess } from '../actions/user';
import { ABSOLUTE_URL_BASE_KEY } from '../reducers/config';

export function configFetch() {
  return {
    type: ActionTypes.CONFIG_FETCH
  };
}

export function configFetchSuccess(config) {
  return {
    type: ActionTypes.CONFIG_FETCH_SUCCESS,
    config
  };
}

export function configFetchError() {
  return {
    type: ActionTypes.CONFIG_FETCH_ERROR
  };
}

export function asyncConfigFetch() {
  return dispatch => {
    dispatch(configFetch());

    let opts = {};
    opts.headers = { Accept: 'application/javascript' };
    http.get('./config.js', opts, function(error, response) {
      if (response) {
        let config = JSON.parse(response.text);
        dispatch(configFetchSuccess(config));
        if (typeof absoluteUrlBase !== 'undefined') {
          /*
                     * Some integrations don't work with relative paths because the URL doesn't match
                     * the actual file path, this function allows integrations to configure a base absolute
                     * url path to be used in components/Image. absoluteBaseUrl should be defined globally
                     * on the page where the SPA is loaded.
                     */
          dispatch(configUpdateByKey(ABSOLUTE_URL_BASE_KEY, absoluteUrlBase));
        }
        dispatch(asyncIntlFetchTranslations(config.locale));
        //log user in if their email is in local storage
        if (isLoggedIn()) {
          dispatch(asyncUserLoginSuccess(getEmail()));
        }
      } else {
        dispatch(configFetchError());
        dispatch(notificationAddError(error));
      }
    });
  };
}

export function configUpdateByKey(key, value) {
  return {
    type: ActionTypes.CONFIG_UPDATE_BY_KEY,
    key,
    value
  };
}
