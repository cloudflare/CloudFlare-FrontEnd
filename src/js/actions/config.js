import http from 'cf-util-http';

import * as ActionTypes from '../constants/ActionTypes';
import { asyncIntlFetchTranslations } from './intl';
import { notificationAddError } from './notifications';
import { isLoggedIn, getEmail } from '../utils/Auth/Auth';
import { asyncUserLoginSuccess } from '../actions/user';

const ABSOLUTE_URL_BASE = "absoluteUrlBase";

export function configFetch() {
    return {
        type: ActionTypes.CONFIG_FETCH
    }
}

export function configFetchSuccess(config) {
    return {
        type: ActionTypes.CONFIG_FETCH_SUCCESS,
        config
    }
}

export function configFetchError() {
    return {
        type: ActionTypes.CONFIG_FETCH_ERROR,
    }
}

export function asyncConfigFetch() {
    return dispatch => {
        dispatch(configFetch());

        let opts = {};
        opts.headers = {'Accept': 'text/javascript'};
        http.get('./config.js', opts, function (response) {
                let config = JSON.parse(response.text);
                config = addAbsoluteUrlBase(config);
                dispatch(configFetchSuccess(config));
                dispatch(asyncIntlFetchTranslations(config.locale))
                //log user in if their email is in local storage
                if(isLoggedIn()) {
                    dispatch(asyncUserLoginSuccess(getEmail()));
                }

            },
            function (error) {
                dispatch(configFetchError());
                dispatch(notificationAddError(error));
            });
    }
}

/*
 * Some integrations don't work with relative paths because the URL doesn't match
 * the actual file path, this function allows integrations to configure a base aboslute
 * url path to be used in components/Image. absoluteBaseUrl should be defined globally
 * on the page where the SPA is loaded.
 */
function addAbsoluteUrlBase(config) {
    if(absoluteUrlBase) {
        config[ABSOLUTE_URL_BASE] = absoluteUrlBase;
    } else {
        config[ABSOLUTE_URL_BASE] = "";
    }
    return config;
}