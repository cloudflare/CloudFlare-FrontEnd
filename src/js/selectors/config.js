import { ABSOLUTE_URL_BASE_KEY } from '../reducers/config';

export function getAbsoluteUrl(config, url) {
    let baseUrl = (typeof config.config[ABSOLUTE_URL_BASE_KEY] !== 'undefined' ? config.config[ABSOLUTE_URL_BASE_KEY] : "");
    return baseUrl + url;
}