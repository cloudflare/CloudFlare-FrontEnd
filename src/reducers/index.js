import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';
import { activeZoneReducer } from './activeZone';
import { appReducer } from './app';
import { configReducer } from './config';
import { dnsRecordsReducer } from './zoneDnsRecords.js';
import { intlReducer } from './intl';
import { notificationsReducer } from './notifications';
import { userReducer } from './user';
import { zoneAnalyticsReducer } from './zoneAnalytics';
import { zonePurgeCacheReducer } from './zonePurgeCache';
import { zoneRailgunReducer } from './zoneRailgun';
import { zoneSettingsReducer } from './zoneSettings';
import { zonesReducer } from './zones';
import { pluginSettingsReducer } from './pluginSettings';

const rootReducer = combineReducers({
    activeZone: activeZoneReducer,
    app: appReducer,
    config: configReducer,
    intl: intlReducer,
    user: userReducer,
    notifications: notificationsReducer,
    routing: routerReducer,
    zones: zonesReducer,
    zoneAnalytics: zoneAnalyticsReducer,
    zoneDnsRecords: dnsRecordsReducer,
    zonePurgeCache: zonePurgeCacheReducer,
    zoneRailguns: zoneRailgunReducer,
    zoneSettings: zoneSettingsReducer,
    pluginSettings: pluginSettingsReducer,
});

export default rootReducer;