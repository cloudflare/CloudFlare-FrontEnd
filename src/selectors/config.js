import { ABSOLUTE_URL_BASE_KEY } from '../reducers/config';

export function getAbsoluteUrl(config, url) {
  let baseUrl = typeof config.config[ABSOLUTE_URL_BASE_KEY] !== 'undefined'
    ? config.config[ABSOLUTE_URL_BASE_KEY]
    : '';
  return baseUrl + url;
}

export function getConfigValue(config, key) {
  return config.config[key];
}

export function isDNSPageEnabled(config) {
  return getConfigValue(config, 'isDNSPageEnabled') === true;
}
