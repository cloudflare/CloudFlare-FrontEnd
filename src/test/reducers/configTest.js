import { configReducer } from '../..//reducers/config';
import * as ActionTypes from '../..//constants/ActionTypes';

describe('Config Reducer', () => {
  it('should return the initial state', () => {
    expect(configReducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle CONFIG_FETCH', () => {
    expect(
      configReducer(undefined, {
        type: ActionTypes.CONFIG_FETCH
      })
    ).toMatchSnapshot();
  });

  it('should handle CONFIG_FETCH_SUCCESS', () => {
    expect(
      configReducer(undefined, {
        type: ActionTypes.CONFIG_FETCH_SUCCESS
      })
    ).toMatchSnapshot();
  });

  it('should handle CONFIG_FETCH_ERROR', () => {
    expect(
      configReducer(undefined, {
        type: ActionTypes.CONFIG_FETCH_ERROR
      })
    ).toMatchSnapshot();
  });

  it('should handle CONFIG_UPDATE_BY_KEY', () => {
    expect(
      configReducer(undefined, {
        type: ActionTypes.CONFIG_UPDATE_BY_KEY,
        key: 'key',
        value: 'value'
      })
    ).toMatchSnapshot();
  });
});
