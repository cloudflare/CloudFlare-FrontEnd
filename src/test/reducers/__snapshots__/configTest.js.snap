// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Reducer should handle CONFIG_FETCH 1`] = `
Object {
  "config": Object {},
  "isFetching": true,
}
`;

exports[`Config Reducer should handle CONFIG_FETCH_ERROR 1`] = `
Object {
  "config": Object {},
  "isFetching": false,
}
`;

exports[`Config Reducer should handle CONFIG_FETCH_SUCCESS 1`] = `
Object {
  "config": Object {},
  "isFetching": false,
}
`;

exports[`Config Reducer should handle CONFIG_UPDATE_BY_KEY 1`] = `
Object {
  "config": Object {
    "key": "value",
  },
  "isFetching": false,
}
`;

exports[`Config Reducer should return the initial state 1`] = `
Object {
  "config": Object {},
  "isFetching": false,
}
`;
